benchmarks/DASB/LibriSpeech/ASR-refactor/train.py

```diff
@@ -0,0 +1,448 @@
+#!/usr/bin/env/python3
+"""Recipe for training an discrete tokens ctc ASR system with librispeech.
+
+Decoding is performed with greedy decoding at validation time.
+At test time, beamsearch is used with an optional external language model.
+
+Authors
+ * Pooneh Mousavi 2024
+"""
+
+import os
+import sys
+import torch
+import torchaudio
+import logging
+import speechbrain as sb
+from speechbrain.utils.distributed import run_on_main, if_main_process
+from speechbrain.tokenizers.SentencePiece import SentencePiece
+from hyperpyyaml import load_hyperpyyaml
+from pathlib import Path
+
+base_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../"))
+sys.path.append(base_dir)
+
+
+logger = logging.getLogger(__name__)
+
+_CACHE = {"size": 0}
+
+
+# Define training procedure
+class ASR(sb.Brain):
+    def compute_forward(self, batch, stage):
+        """Forward computations from the waveform batches to the output probabilities."""
+        batch = batch.to(self.device)
+        wavs, wav_lens = batch.sig
+
+        # Add waveform augmentation if specified.
+        if stage == sb.Stage.TRAIN and hasattr(self.hparams, "wav_augment"):
+            wavs, wav_lens = self.hparams.wav_augment(wavs, wav_lens)  # [B, T]
+
+        # compute features
+        # Extract tokens (cache them at first epoch if augmentation is disabled)
+        key = tuple(sorted(batch.id))
```

Comment:
This caching code works well only if batch_size=1 or batch_size > 1 and sorting is not random.